Devs/jptiz/optional ector

JPTIZ opened the pull request.

JPTIZ committed:
build: add option to enable/disable ector

JPTIZ committed:
evas: search for filters only when ector is enabled

JPTIZ committed:
evas: include vg sources only when ector is enabled

JPTIZ committed:
evas: isolate ector parts of software_generic when ector is disabled

JPTIZ committed:
evas module: ignore filters when ector is not enabled

JPTIZ committed:
evas: ignore filters when ector is not enabled - but for software_generic now

JPTIZ committed:
evas: placing a lot of #ifdef HAVE_ECTOR

JPTIZ committed:
evas module: placing a lot of #ifdef HAVE_ECTOR

felipealmeida committed:
ector: Remove compilation of VG files and some ifdefs

JPTIZ committed:
edje: enable edje "vector" part only if ector is enabled

edje: split vector api into its own file
edje: ignore vector mempool deletions

JPTIZ committed:
edje: ignore filter feature test

JPTIZ committed:
Skip some failing tests

JPTIZ committed:
elementary: add HAVE_ECTOR condition to generate Efl.Ui.Vg_Animation

JPTIZ committed:
evas: fix evas_text_object pad getter when -Dector=false

JPTIZ committed:
evas canvas: allow render unref if ector

JPTIZ committed:
evas: split vg into its own legacy header

JPTIZ committed:
edje: do not write vectors if ector is disabled

JPTIZ committed:
tests: disable vg/filters tests and examples when ector is disabled

JPTIZ committed:
evas module: isolate ector and filters when ector is disabled

JPTIZ committed:
elementary: disable vg when ector is disabled

JPTIZ committed:
evas: Isolate internal filter

JPTIZ committed:
edje: ignore vg functions

JPTIZ committed:
efl interfaces: remove text_gfx_filter from Efl.Text_Style

JPTIZ committed:
evas canvas: remove has_filter if no ector

JPTIZ committed:
(?) evas: tons of have_ector

JPTIZ committed:
Remove Efl.Ui.Vg_Animation

Coquinho commented on meson.build:
```diff
@@ -498,6 +498,7 @@ foreach package : subprojects
       endif
     endif
   else
+    set_variable(package_name, declare_dependency())
```
Sorry, but I didn't get why this is needed, could you explain for me?

JPTIZ replied:
At first, some meson.builds were complaining about unknown static_libs, so this was just a shortcut to get it compiling (no preprocessor symbol is defined, so `HAVE_*` won't mislead, etc.).

Coquinho commented on src/lib/edje/edje_calc.c:
```diff
@@ -3262,7 +3271,6 @@ _edje_vector_load_lottie(Edje *ed, Edje_Real_Part *ep, const char *key)
         efl_gfx_frame_controller_frame_set(ep->object, (int)(frame_count * desc->vg.frame));
      }
 }
-#endif
```
Why changing where the `endif` of  `#ifdef BUILD_VG_LOADER_JSON`?

JPTIZ replied:
Because it makes not much sense to have vector functions if vg_loader is disabled, I think, but looking at the code I think this was a misunderstanding.

Coquinho replied:
For me, it seems like a misunderstanding too, I mean, It was there for a reason, and if it was wrong or not shouldn't be decided at this PR.

JPTIZ replied:
I'll address that in the next commit.

JPTIZ replied:
Done.

Coquinho commented on src/lib/elementary/meson.build:
```diff
@@ -949,6 +963,10 @@ elementary_src = files([
   'efl_ui_spotlight_animation_manager.c',
 ])
 
+if get_option('ector')
+  elementary_src += 'efl_ui_vg_animation.c'
+endif
```
Does it work? Shouldn't it be:
```suggestion
if get_option('ector')
  elementary_src += files(['efl_ui_vg_animation.c'])
endif
```

Coquinho commented on src/lib/evas/canvas/efl_canvas_textblock.eo:
```diff
@@ -316,7 +316,7 @@ class Efl.Canvas.Textblock extends Efl.Canvas.Object implements Efl.Text,
              See @Efl.Text_Style.text_underline_height.
 
-           - $gfx_filter: Experimental filter name. See @Efl.Gfx.Filter for more information.
+           - gfx_filter: Experimental filter name. See Efl.Gfx.Filter for more information.
```
What `$` and `@` do here? the first one is to see, `gtx_filter` as a variable a,d the second to set a link to `Efl.Gfx.Filter`?

JPTIZ replied:
@Coquinho they're referencing other Eo classes. What I do here is prevent the reference, so the doc generator doesn't try to look for them.

Coquinho replied:
Oh, ok.
And why do we need to change docs here?

JPTIZ replied:
Because otherwise it won't compile, since if the doc parser doesn't find the references, it gives a compilation error.

Coquinho commented on src/lib/evas/canvas/evas_object_text.c:
```diff
@@ -675,10 +681,13 @@ _evas_object_text_pad_get(const Eo *eo_obj, Evas_Text_Data *o, int *l, int *r, i
+#ifdef HAVE_ECTOR
+   if (o->has_filter)
      efl_gfx_filter_padding_get(eo_obj, l, r, t, b);
+   else
+#else
```
There is no `else` block here Oo

JPTIZ replied:
Oh true

Coquinho commented on src/lib/evas/canvas/evas_object_textblock.c:
```diff
@@ -342,7 +349,9 @@ struct _Efl_Canvas_Textblock_Filter
    Evas_Object          *eo_obj;
    Evas_Public_Data     *evas;
    void                 *dc; /* draw context - no clip, white, no colmul... */
+#ifdef HAVE_ECTOR
    Evas_Filter_Padding   pad;
+#endif
```
I think you don't need an `#ifdef HAVE_ECTOR` inside an `#ifdef HAVE_ECTOR`

JPTIZ commented:
@Coquinho I addressed some of your comments.

Coquinho commented on src/lib/evas/meson.build:
```diff
@@ -134,16 +141,25 @@ evas_include_directories += [
 ]
 
+if get_option('ector')
+  evas_include_directories += [
+    include_directories('filters'),
+  ]
+endif
+
 evas_src += files([
    'main.c'
 ])
 
 evas_src_opt = [ ]
 
-evas_ext_none_static_deps += dependency('freetype2')
+if get_variable('freetype', false) == false
+  freetype = dependency('freetype2')
+endif
+
+evas_ext_none_static_deps += freetype
```
What was the problem with FreeType?

JPTIZ replied:
Tbh I find this strange as well. I think I applied an extra `git stash pop` by accident.

JPTIZ replied:
Oooh right right, now I remember. This is sort of a fix to the `freetype` dependency. First, see the first appearence of `freetype` in `src/lib/ector/meson.build`:

```meson
ector_ext_deps = [triangulator, freetype, draw_without_deps, m]
```

It gets freetype from the static_lib, see the main `/meson.build`:

```meson
  ['ector'            ,['ector']             , false,  true, false,  false,   true,     false,  true, ['eina', 'efl'], ['draw', 'triangulator', 'freetype']],
```

Which defines the `freetype` variable and etc. What I did was to always get the variable when it exists, otherwise get from the system-wide dependency.

Coquinho replied:
Oh ok, make sense.
But wouldn't be better to do this at the `meson.build` at the root of the project?

Coquinho replied:
Oh, just noticed that `freetype` is declared at `src/static_libs/freetype/meson.build` so maybe there would be even better. What do you think?

JPTIZ replied:
It could be, yes.

Coquinho commented on src/modules/evas/engines/gl_generic/meson.build:
```diff
@@ -45,10 +51,12 @@ foreach src : common_engine_src
-  'evas_ector_gl_buffer.eo',
-  'evas_ector_gl_image_buffer.eo'
-]
+if get_option('ector')
+  pub_eo_files = [
+    'evas_ector_gl_buffer.eo',
+    'evas_ector_gl_image_buffer.eo'
+  ]
+endif
```
Shouldn't `pub_eo_files` be declared empty for the case where there is no ector?
```suggestion
pub_eo_files = []
if get_option('ector')
  pub_eo_files += [
    'evas_ector_gl_buffer.eo',
    'evas_ector_gl_image_buffer.eo'
  ]
endif
```

Or else, the  `foreach` at line61 will yell that there are no `pub_eo_files`, right?

JPTIZ replied:
Yes, or worse: it may get a previously defined `pub_eo_files`.

Coquinho commented on src/modules/evas/engines/software_generic/filters/evas_filter_bump.c:
```diff
@@ -103,9 +103,13 @@ _bump_map_cpu_alpha_alpha(Evas_Filter_Command *cmd)
    h = cmd->input->h;
    EINA_SAFETY_ON_FALSE_RETURN_VAL(w > 2 && h > 2, EINA_FALSE);
 
+#define E_READ 0
+#define E_WRITE 0
```
Shouldn't this be defined elsewhere (probably at `src/lib/evas/filters/evas_filter_private.h:310` as an `#else` to the `#ifdef HAVE_ECTOR` at line 308) ? Like this we will overwrite any previous definition, won't we?

JPTIZ replied:
Actually these files aren't even compiled without ector, hm...

Coquinho commented on src/modules/evas/engines/software_generic/filters/evas_filter_grayscale.c:
```diff
@@ -1,4 +1,6 @@
 #include "evas_engine_filter.h"
+#define E_READ 0
+#define E_WRITE 0
```
Same as https://github.com/expertisesolutions/efl/pull/478/files#r523233109

Coquinho reviewed (commented):
Do files at `src/lib/evas/software_generic/filters/` need any changes? `src/lib/evas/software_generic/filters/meson.build` has an `subdir_done` in case ector is disabled, and `src/lib/evas/software_generic/meson.build` don't even add `filter` as a subdir in that case.

Coquinho commented on src/tests/elementary/spec/efl_test_gfx_view.c:
```diff
@@ -6,10 +6,12 @@
 
+#ifdef HAVE_ECTOR
+#else
 /* spec-meta-start
-      {"test-interface":"Efl.Gfx.View",
-       "test-widgets": ["Efl.Ui.Vg_Animation"]}
+      {"test-interface":"Efl.Gfx.View"}
    spec-meta-end */
+#endif
```
Is this as intended? It seems as something went wrong here, or at least it should be a `#ifndef` instead of a `#ifdef`

JPTIZ replied:
The `#ifdef` here actually makes no changes at all to the comments. This is a leftover of an attempt to catch that behaviour.

JPTIZ committed:
Address @Coquinho's comments

- Removed HAVE_ECTOR for files that are not compiled.
- Fixed pub_eo_files definition in gl_generic not being empty when ector
  is not defined.
- Removed unnecessary HAVE_ECTOR in efl_test_gfx_view and bring back
  Vg_Animation object from spec JSON comment.
- Undo `#endif` change in `#ifdef BUILD_VG_LOADER_JSON`.

Coquinho commented on src/tests/elementary/elm_test_check.c:
```diff
@@ -196,10 +198,12 @@ EFL_START_TEST(elm_test_check_api_call)
    click_object(data.check1);
    ecore_main_loop_iterate();
-   ck_assert_int_eq(data.value, 1);
+   ck_assert_int_eq(data.value, 1); // FIXME No more ifdef have_ector
```
What's the problem here?

JPTIZ replied:
Problem is events' callbacks aren't being called for Eo objects, so that `data.value` is never set.

Coquinho committed:
Evas: Remove need for duplicated code at \#ifdef HAVE_ECTOR (#479)

Coquinho committed:
Evas: Resolve some undefined references to Ector + Evas types (#480)

As `src/lib/evas/include/evas_filter.h` don't include `config.h`,
`HAVE_ECTOR` wasn't defined. Which caused almost all the file to be
ignored, making some types defined there unavailable, causing some
undefined references.

This resolve that problem by removing `evas_common_private.h` and
`evas_private.h` from the guard, as they already have theirs, they don't
need to be inside one) which do include `config.h`, making `HAVE_ECTOR`
available when `-Dector=true`.

JPTIZ committed:
build: do not set empty dependency variables

JPTIZ committed:
edje: pass *all* vector functions to _vg file

JPTIZ committed:
have a padding when filters (ector) is disabled

JPTIZ committed:
small indent fix

JPTIZ committed:
wip wip wip

JPTIZ committed:
evas canvas: delete lines with filters instead of comment

JPTIZ committed:
[*] Do not change files that are not compiled when ector=false

JPTIZ committed:
[*] Do not throw the dirt under the rug

JPTIZ committed:
evas canvas: add a reason to set parameter as unused

JPTIZ committed:
[*] Small attempt to confirm it wasn't an issue with unitialized variable

JPTIZ committed:
[*] Missed one ector eo file

JPTIZ committed:
[*] Do not change evas_filter.h since it won't be included at all.

JPTIZ committed:
vg_common: do not build if ector is disabled

JPTIZ committed:
ci: show meson commands before executing them
